AVSS refactoring

smkuls opened the pull request:
**Change missing randints to field.random()**

**Cleanup AVSS code**
* Move redundant polynomial code from AVSS code and reuse code in
`polynomial.py`. This required modifying some code in `betterpairing.py`
in the `ZR` class to support certain operations.
* Move symmetric cryptography code out from AVSS to a separate file
and add tests for it.
* Move polynomial commitment code out from AVSS to a separate file and
add tests for it.
* Change `ZR.rand()` to `ZR.random()` to be consistent with
`field.random()` so both can be used seamlessly in the  `Polynomial`
class.

**Reimplementing HbAVSS light**
HbAVSS light has been reimplemented for the purpose of offline phase.
This also does not assume that the dealer is outside of the parties that are
participating which is the case in the offline phase. The client mode is
also supported where the dealer is not one of the parties
participating in the MPC as was the case with the previous version.

smkuls committed:
Change missing randints to field.random()

smkuls committed:
Cleanup AVSS code

* Move redundant polynomial code from AVSS code and reuse code in
`polynomial.py`. This required modifying some code in `betterpairing.py`
in the `ZR` class to support certain operations.
* Move symmetric cryptography code out from AVSS to a separate file
and add tests for it.
* Move polynomial commitment code out from AVSS to a separate file and
add tests for it.
* Change `ZR.rand()` to `ZR.random()` to be consistent with
`field.random()` so both can be used seamlessly in the  `Polynomial`
class.

smkuls committed:
Reimplementing HbAVSS light

HbAVSS light has been reimplemented for the purpose of offline phase.
This also does not assume that the dealer is outside of the parties that are
participating which is the case in the offline phase. The client mode is
also supported where the dealer is not one of the parties
participating in the MPC as was the case with the previous version.

amiller commented on honeybadgermpc/hbavss_light.py:
```diff
@@ -0,0 +1,135 @@
+    OK = "OK"
+
+
+class HbAvss(object):
```
name this HbAvss light to make room for the batch version?

smkuls replied:
Done.

amiller commented on honeybadgermpc/hbavss_light.py:
```diff
@@ -0,0 +1,135 @@
+        # When running multiple AVSSes, the messages are forwarded to the correct
+        # AVSS using the queue. The key is the AVSS ID which is sent along with
+        # each AVSS message.
+        self.queues = defaultdict(asyncio.Queue)
```
So a single instance of this `HbAvss` object supports multiple AVSS's from the same dealer, right?

smkuls replied:
Yes, this is demonstrated in the tests as well.

amiller commented on honeybadgermpc/hbavss_light.py:
```diff
@@ -0,0 +1,135 @@
+            self.t,
+            dealer_id,
+            broadcast_msg,
+            self.queues[avss_id].get,
```
There is a potential message ordering problem here. You are using the same read channel `self.queues[avss_id].get` for both the reliable broadcast subprotocol and for the AVSS-specific OK messages. It is possible that these two kinds of messages can be interleaved. These will need to be partitioned somehow into, for example, `self.RBCqueues` and `self.AVSSqueues` based on a tag

smkuls replied:
Makes sense. If someone finishes RBC and starts sending AVSS OK messages then that will screw things up. Will split this into avss and rbc queues.

smkuls replied:
This has been fixed.

amiller reviewed (changes requested):
see note about potential race condition

amiller commented on honeybadgermpc/hbavss_multi.py:
```diff
@@ -249,24 +52,20 @@ def __init__(self, publicparams, privateparams, send, recv):
-        sk = ZR.rand()
+        (sharedkeys, shares, encryptedshares,
+         witnesses, encryptedwitnesses) = ({}, {}, {}, {}, {})
+        poly = polynomials_over(use_rust=True).random(t, secret)
```
doesn't `polynomials_over` have a required first argument?

smkuls replied:
I have removed it. I have also added an `avss_parallel` method in `HbAvssLight` which effectively replaces `hbavss_multi.py` and added a test case for the same.

This is also how I expect the batch interface to be.

amiller reviewed (changes requested):
I think what's happening is that there are no test cases for `hbavss_multi.py`, yet hbavss_multi was also changed. As part of this refactoring we should probably get rid of `hbavss_multi` or at least refactor it not to be redundant

smkuls committed:
Implement AVSS parallel and remove hbavss_multi.py

Implement a parallel version of hbavss light to share multiple values at
one time. This was achieved by hbavss_multi.py earlier which is also
removed in this commit.

smkuls committed:
Remove AVSS related multi processing code

The code is outdated and can be added later on when needed. It doesn't
work because of the AVSS code refactoring.
